### PR TITLE
feat: added content-digest header & generation logic for gnap requests

### LINKED ISSUE #368: Implement content-digest header generation
As per [GNAP spec](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-09.html#section-7.3.1-5.2.1), requests need to have a `content-digest` header generated for each request.

### PR DESCRIPTION
**Pull request by birtony**
Fixes #368

Added `content-digest` headers to GNAP requests and corresponding logic as per the [specs](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-09.html#section-7.3.1-2).

Signed-off-by: Anton Biriukov <anton.biriukov@securekey.com>

### CODE REVIEW

File: cmd/wallet-js-sdk/src/gnap/client.js [hunk 1]
```diff
@@ -57,19 +60,18 @@ export class Client {
...
-          // "Signature-Input": "TODO", // TODO update signature input
-          // TODO : Implement http message signature module in JS (https://github.com/trustbloc/agent-sdk/issues/348)
-          // Signature: encode(sig, true),
-        },
-      }
-    );
+    const gnapResp = await axios.post(url, req, {
+      headers: {
+        "Content-Type": CONTENT_TYPE,
+        "@target-uri": url,
```
[the remark this answers is not in the document]
Comment (Baha-sk) [reply cmd/wallet-js-sdk/src/gnap/client.js:68]: no, see example `@target-uri` further down in this section: https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-09.html#section-7.3.1

File: cmd/wallet-js-sdk/src/gnap/client.js [hunk 2]
```diff
@@ -57,19 +60,18 @@ export class Client {
...
+        "content-digest": encode(getDigest("SHA-256", req)),
```
Comment (Baha-sk) [inline cmd/wallet-js-sdk/src/gnap/client.js:69]: add a TODO to support `SHA-512` as well

File: cmd/wallet-js-sdk/src/httpsig/digest.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * Generates and returns a hash for the data object provided using the specified encryption algorithm.
+ *
+ *  @param {String} name - name of the encryption algorithm to be used for generating the digest.
+ *  @param {Object} data - data object to generate the digest for.
+ *
+ * @returns {String} - generated hash for the data provided
+ */
+export default async function getDigest(name, data) {
+  switch (name) {
+    case "SHA-256":
```
Comment (Baha-sk) [inline cmd/wallet-js-sdk/src/httpsig/digest.js:17]: support `SHA-512` as well

File: cmd/wallet-js-sdk/test/specs/httpsig/digest.spec.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    client: {
+      key: {
+        proof: "httpsig",
+        jwk: {
+          kty: "EC",
+          kid: "key1",
+          crv: "P-256",
+          alg: "ES256",
+          x: "TDiE9hqocEOewpSgI5r_yhoOy1EuetUpWcKvlrC9Ix4",
+          y: "pC76IX_6YRQ5CaxmcmcixTA-dWjHpuKoYICir7VYQ_A",
```
Comment (Baha-sk) [inline cmd/wallet-js-sdk/test/specs/httpsig/digest.spec.js:64]: make these variables to allow different key types ( can be done in a future enhancement)

**Review by Baha-sk — APPROVED** [review]
looks good, added some nitpcks that can be done in future enhancements

File: cmd/wallet-js-sdk/src/gnap/client.js (hunk 2 above)
Comment (birtony) [reply cmd/wallet-js-sdk/src/gnap/client.js]: Added `SHA-512` support

File: cmd/wallet-js-sdk/test/specs/httpsig/digest.spec.js (hunk 4 above)
Comment (birtony) [reply cmd/wallet-js-sdk/test/specs/httpsig/digest.spec.js]: Updated

File: cmd/wallet-js-sdk/src/gnap/client.js [hunk 5]
```diff
@@ -6,7 +6,8 @@ SPDX-License-Identifier: Apache-2.0
 
 import axios from "axios";
 // TODO : Implement http message signature module in JS (https://github.com/trustbloc/agent-sdk/issues/348)
-// import { encode } from "js-base64";
+// import { encodeURI } from "js-base64";
```
Comment (Moopli) [inline cmd/wallet-js-sdk/src/gnap/client.js:9]: commented-out

File: cmd/wallet-js-sdk/src/gnap/client.js (hunk 1 above)
Comment (Moopli) [inline cmd/wallet-js-sdk/src/gnap/client.js:68]: target uri doesn't need to be a header - the http-signature algorithm is meant to grab the target URI of the request itself.

File: cmd/wallet-js-sdk/src/gnap/client.js [hunk 6]
```diff
@@ -44,6 +45,8 @@ export class Client {
     const gnapResp = await axios.post(url, req, {
       headers: {
         "Content-Type": CONTENT_TYPE,
+        "@target-uri": url,
```
Comment (Moopli) [inline cmd/wallet-js-sdk/src/gnap/client.js:48]: target uri doesn't need to be a header - the http-signature algorithm is meant to grab the target URI of the request itself.

File: cmd/wallet-js-sdk/src/gnap/client.js (hunk 5 above)
Comment (birtony) [reply cmd/wallet-js-sdk/src/gnap/client.js:9]: This was commented out before, it's not needed for now - it's a part of the TODO above. I just noticed we need to use `encodeURI` instead, so I updated it to not miss it later

**Review by Moopli — APPROVED** [review]

**Pull request merged**